src/state/index.tsx

```diff
@@ -1,52 +1,67 @@
-import React, { createContext, useCallback, useContext, useState } from 'react';
+import React, { createContext, useContext, useState } from 'react';
 import { TwilioError } from 'twilio-video';
-import useAuth from './useAuth/useAuth';
+import useFirebaseAuth from './useFirebaseAuth/useFirebaseAuth';
+import usePasscodeAuth from './usePasscodeAuth/usePasscodeAuth';
 import { User } from 'firebase';
 
 export interface StateContextType {
   error: TwilioError | null;
   setError(error: TwilioError | null): void;
-  getToken(name: string, room: string): Promise<string>;
-  user: User | null;
-  signIn(): Promise<void>;
-  signOut(): Promise<void>;
-  isAuthReady: boolean;
+  getToken(name: string, room: string, passcode?: string): Promise<string>;
+  user?: User | null | { displayName: undefined; photoURL: undefined; passcode?: string };
+  signIn?(passcode?: string): Promise<void>;
+  signOut?(): Promise<void>;
```

Comment:
It seems like we're going to be doing optional chaining whenever we use signIn and signOut anywhere.  Is there a way to handle this inside the function so we don't have to do `signOut?()` or `signIn?()` everywhere?

Reply:
I think `signIn?.()` is appropriate, because if `process.env.REACT_APP_SET_AUTH` is undefined (in the case that the user is using the local token server), then there actually are no signIn or signOut functions.

However, the UI elements that call the signIn and signOut functions are not even rendered when `REACT_APP_SET_AUTH` is undefined. So this may be a case where we can use `signIn!()` as a way to tell TypeScript "trust me, signIn will definitely be a function". 

I'm still leaning towards optional chaining though. I'm not sure it makes a big difference either way - it just feels a little safer to use. Without it I'd be worried about `signIn is not a function` errors. It also feels more descriptive. Those functions _can_ be undefined. 

What do you think?

Reply:
I also think optional chaining sounds better over `signIn!()`. I was just concerned this might get used everywhere and see a bunch of optional chaining. Though not always bad, but usually a code smell if there are too many. I'm ok to keep it as is with the current use case.